File: medium/setZeroMatrix.py
class Solution(object):
    def setZeroes(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: None Do not return anything, modify matrix in-place instead.
        """
        m, n = len(matrix), len(matrix[0])
        col_set = set()
        row_set = set()
        
        # First pass to find all the zeros
        for i in range(m):
            for j in range(n):
                if matrix[i][j] == 0:
                    col_set.add(j)
                    row_set.add(i)
        
        # Zero out the rows
        for i in row_set:
            for j in range(n):
                matrix[i][j] = 0
        
        # Zero out the columns
        for j in col_set:
            for i in range(m):
                matrix[i][j] = 0

File: medium/setZeroMatrix.cpp
#include <iostream>
#include <set>

//more efficient implementation of set_zero_matrix.py. imo a more efficient solution (less memory complexity) prolly involves just using a boolean but i'm lazy lol
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int rows = matrix.size();
        int cols = matrix[0].size();

        std::set<int> row_set;
        std::set<int> col_set;

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (matrix[r][c] == 0) {
                    row_set.insert(r);
                    col_set.insert(c);
                }
            }
        }

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (row_set.count(r) != 0 || col_set.count(c) != 0) {
                    matrix[r][c] = 0;
                }
            }
        }
    }
};
